pallets/nodle-staking/src/lib.rs

```diff
@@ -0,0 +1,2311 @@
+        /// See [Era payout](./index.html#era-payout).
+        type RewardRemainder: OnUnbalanced<NegativeImbalanceOf<Self>>;
+        /// Interface for interacting with a session module.
+        type SessionInterface: self::SessionInterface<Self::AccountId>;
+        /// This pallet's module id. Used to derivate a dedicated account id to store session
+        /// rewards for validators and nominators in.
+        type PalletId: Get<ModuleId>;
+        /// The origin which can cancel a deferred slash. Root can always do this.
+        type CancelOrigin: EnsureOrigin<Self::Origin>;
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(crate) trait Store)]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Set the validators who cannot be slashed (if any).
+        ///
+        /// The dispatch origin must be Root.
+        #[pallet::weight(T::WeightInfo::set_invulnerables(invulnerables.len() as u32))]
+        pub fn set_invulnerables(
+            origin: OriginFor<T>,
+            invulnerables: Vec<T::AccountId>,
+        ) -> DispatchResultWithPostInfo {
+            T::CancelOrigin::try_origin(origin)
+                .map(|_| ())
+                .or_else(ensure_root)?;
+            <Invulnerables<T>>::put(&invulnerables);
+            Self::deposit_event(Event::NewInvulnerables(invulnerables));
+            Ok(().into())
+        }
+        /// Set the total number of validator selected per round
+        /// - changes are not applied until the start of the next round
+        #[pallet::weight(T::WeightInfo::set_total_validator_per_round(*new))]
+        pub fn set_total_validator_per_round(
+            origin: OriginFor<T>,
+            new: u32,
+        ) -> DispatchResultWithPostInfo {
+            T::CancelOrigin::try_origin(origin)
+                .map(|_| ())
+                .or_else(ensure_root)?;
+            ensure!(
+                new >= T::MinSelectedValidators::get(),
+                Error::<T>::CannotSetBelowMin
+            );
+            let old = <TotalSelected<T>>::get();
+            <TotalSelected<T>>::put(new);
+            Self::deposit_event(Event::TotalSelectedSet(old, new));
+            Ok(().into())
+        }
+        /// Join the set of validators pool
+        #[pallet::weight(T::WeightInfo::validator_join_pool())]
+        pub fn validator_join_pool(
+            origin: OriginFor<T>,
+            bond: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            log::debug!("validator_join_pool:[{:#?}] - Entry!!!", line!(),);
+
+            let acc = ensure_signed(origin)?;
+
+            ensure!(!Self::is_validator(&acc), Error::<T>::ValidatorExists);
+            ensure!(!Self::is_nominator(&acc), Error::<T>::NominatorExists);
```

Comment:
Are there any benefits to forbidding nominators from becoming validators?

Reply:
:+1:  Done